tensorboard/webapp/core/store/core_selectors.ts

```diff
@@ -35,13 +36,74 @@ export const getPluginsListLoaded = createSelector(
   (state: CoreState): PluginsListLoadState => state.pluginsListLoaded
 );
 
-// TODO(tensorboard-team): AppLastLoaded is currently derived from plugins listing loaded
-// state which should be disentangled. Fix this by having a separate state for remembering
-// when the application data was last loaded.
+export const getPolymerRunsLoadState = createSelector(
+  selectCoreState,
+  (state: CoreState): LoadState => state.polymerRunsLoadState
+);
+
+export const getCoreDataLoadedState = createSelector(
+  getPluginsListLoaded,
+  getPolymerRunsLoadState,
+  ({state: pluginsLoadState}, {state: runsLoadState}): DataLoadState => {
+    if (pluginsLoadState === runsLoadState) {
+      return pluginsLoadState;
+    }
+    if (
+      pluginsLoadState === DataLoadState.LOADING ||
+      runsLoadState === DataLoadState.LOADING
+    ) {
+      return DataLoadState.LOADING;
+    }
+    if (
+      pluginsLoadState === DataLoadState.FAILED ||
+      runsLoadState === DataLoadState.FAILED
+    ) {
+      return DataLoadState.FAILED;
+    }
+    // State is LOADED only when both states are `LOADED` which is handled by
+    // the first if statement.
+    return DataLoadState.NOT_LOADED;
+  }
+);
+
+// TODO(tensorboard-team): AppLastLoaded is currently derived from only plugins
+// listing loaded state and runs load state. It should be broken down further
+// (such as the environments) and derive this state from the combination of
+// other core data load state.
+/**
+ * Returns last _successful_ loaded time of the applicational state where an
+ * applicational state is defined as combinations of plugins listing, runs, and
+ * environment.
+ *
+ * When `getCoreDataLoadedState` returns LOADING, `getAppLastLoadedTimeInMs`
+ * returns loaded time for the minimum timestamp of all load states while, when
+ * both are loaded, it returns the max of all last loaded time.
+ */
 export const getAppLastLoadedTimeInMs = createSelector(
   getPluginsListLoaded,
-  (loadedState: PluginsListLoadState): number | null =>
-    loadedState.lastLoadedTimeInMs
+  getPolymerRunsLoadState,
+  (
+    pluginListLoadState: PluginsListLoadState,
+    runsLoadState: LoadState
+  ): number | null => {
+    if (
+      pluginListLoadState.lastLoadedTimeInMs === null ||
+      runsLoadState.lastLoadedTimeInMs === null
+    ) {
+      return null;
+    }
+    if (pluginListLoadState.state === runsLoadState.state) {
```

Comment:
There seem to be scenarios where the value returned by getAppLastLoadedTimeInMs can go backwards in time during a reload, which I find really confusing.

I think if you want to solve the problem accurately then it needs to be modeled in the state with a new `coreLoadState: LoadState` property and handled in the reducer layer.

Would this work?

```
on(pluginListFetchRequested):
  pluginListLoadState = {...pluginListLoadState, state: LOADING}
  coreLoadState = {...coreLoadState, state: LOADING}
on(polymerRunsFetchRequested):
  polymerRunsLoadState = {...polymerRunsLoadState, state: LOADING}
  coreLoadState = {...coreLoadState, state: LOADING}
on(pluginListFetchSuccess):
  pluginListLoadState = {state: LOADED, lastLoadTime: Date.now()}
  if (pluginLoadState.state === LOADED && polymerRunsLoadState.state === LOADED):
    coreLoadState = {state: LOADED, lastLoadTime: pluginListLoad.lastLoadTime}
on(polymerRunsFetchSuccess):
  polymerRunsLoadState = {state: LOADED, lastLoadTime: Date.now()}
  if (pluginLoadState.state === LOADED && polymerRunsLoadState.state === LOADED):
    coreLoadState = {state: LOADED, lastLoadTime: polymerRunsLoadState.lastLoadTime}
// TODO: Handle failed states
```

Then getCoreDataLoadedState and getAppLastLoadedTimeInMs become trivial.

Reply:
Thanks for catching the error in my logic. Adopted your change.